docs/topics/db/instrumentation.txt

```diff
@@ -0,0 +1,118 @@
+========================
+Database instrumentation
+========================
+
+.. versionadded:: 2.0
+
+In order to help you understand and control the queries issued by your code,
+Django provides a mechanism for installing wrappers around the execution of
+database queries. Such wrappers can be used to count queries, measure the time
+they take, log them, or even prevent their execution (e.g. to make sure that no
+new database queries are issued while rendering a template, when you expect it
+to use only data which has been fetched already).
+
+The wrappers are modeled after :doc:`middleware </topics/http/middleware>` --
+they are callables which take, as one of their arguments, another callable;
+they call it to invoke the (possibly wrapped) database query, and they can do
+what they will around it. They are, however, created and installed by user
+code, and so do not need a separate factory like middleware do.
+
+Installing a wrapper is done in a context manager -- so the wrappers are
+temporary and can be made specific to some flow in your code.
+
+Probably the most trivial example of a wrapper is an execution blocker as
+mentioned above. It could look like this::
+
+    def blocker(*args):
+        raise Exception("No database access allowed here")
+
+And it would be used in a view to block queries from the template like so::
+
+    from django.db import connection
+    from django.shortcuts import render
+
+    def my_view(request):
+        context = ...  # Code to generate context with all data
+        template_name = ...
+        with connection.execute_wrapper(blocker):
+            return render(request, template_name, context)
+
+The parameters sent to the wrappers are:
+
+* ``execute`` -- a callable, which should be invoked with the rest of the
+  parameters in order to execute the query;
```

Comment:
I think the semicolons are best replaced as full stops / periods.

Reply:
Accepted.